java/serve/src/main/java/io/ray/serve/config/DeploymentConfig.java

```diff
@@ -0,0 +1,263 @@
+package io.ray.serve.config;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.ray.runtime.serializer.MessagePackSerializer;
+import io.ray.serve.common.Constants;
+import io.ray.serve.exception.RayServeException;
+import io.ray.serve.generated.DeploymentLanguage;
+import io.ray.serve.util.LogUtil;
+import java.io.Serializable;
+
+/** Configuration options for a deployment, to be set by the user. */
+public class DeploymentConfig implements Serializable {
+
+  private static final long serialVersionUID = 5965977837248820843L;
+
+  /**
+   * The number of processes to start up that will handle requests to this deployment. Defaults to
+   * 1.
+   */
+  private Integer numReplicas = 1;
+
+  /**
+   * The maximum number of queries that will be sent to a replica of this deployment without
+   * receiving a response. Defaults to 100.
+   */
+  private Integer maxConcurrentQueries = 100;
+
+  /**
+   * Arguments to pass to the reconfigure method of the deployment. The reconfigure method is called
+   * if user_config is not None.
+   */
+  private Object userConfig;
+
+  /**
+   * Duration that deployment replicas will wait until there is no more work to be done before
+   * shutting down.
+   */
+  private Double gracefulShutdownWaitLoopS = Constants.DEFAULT_GRACEFUL_SHUTDOWN_WAIT_LOOP_S;
+
+  /** Controller waits for this duration to forcefully kill the replica for shutdown. */
+  private Double gracefulShutdownTimeoutS = Constants.DEFAULT_GRACEFUL_SHUTDOWN_TIMEOUT_S;
+
+  /** Frequency at which the controller will health check replicas. */
+  private Double healthCheckPeriodS = Constants.DEFAULT_HEALTH_CHECK_PERIOD_S;
+
+  /**
+   * Timeout that the controller will wait for a response from the replica's health check before
+   * marking it unhealthy.
+   */
+  private Double healthCheckTimeoutS = Constants.DEFAULT_HEALTH_CHECK_TIMEOUT_S;
+
+  private AutoscalingConfig autoscalingConfig;
+
+  /** This flag is used to let replica know they are deplyed from a different language. */
+  private Boolean isCrossLanguage = false;
+
+  /** This flag is used to let controller know which language does the deploymnent use. */
+  private DeploymentLanguage deploymentLanguage = DeploymentLanguage.JAVA;
+
+  private String version;
+
+  private String prevVersion;
+
+  private DeploymentLanguage apiLanguage = DeploymentLanguage.JAVA;
```

Comment:
What's the difference between this and `deploymentLanguage`?

Reply:
Resolved. use deployment_language and is_cross_language instead api_language.
We have four scenarios:
1. Deploy java deployment with python serve
2. Deploy python deployment with python serve
3. Deploy java deployment with java serve
4. Deploy python deployment with java serve
The first two need to be serialized using pickle, and the latter two are serialized using msgpack. we cannot use deployment_language to distinguish the serialization type. So api_language was added. But if only have python and java serve, we can use deployment_language and is_cross_language instead api_language now.